### PR TITLE
Update docs to let users know the concurrency problem on SQLite3

### LINKED ISSUE #4002: enqueue, storage & parallelization
### Expected behavior

run the attached code, you will see the problem. this issue does not always happen, but often.

### Environment

- Optuna version: 3.9
- Python version:3.0
- OS: windows X64
- (Optional) Other libraries and their versions:

### Error messages, stack traces, or logs

```shell
[I 2022-09-14 14:22:16,229] A new study created in RDB with name: C:/Users/cuism/Desktop/ex_DOEs/study
A new study created in RDB with name: C:/Users/cuism/Desktop/ex_DOEs/study
trial index istrial index is  00

[I 2022-09-14 14:22:17,167] Trial 0 finished with value: 450.0 and parameters: {'x': -15, 'y': -15}. Best is trial 0 with value: 450.0.
Trial 0 finished with value: 450.0 and parameters: {'x': -15, 'y': -15}. Best is trial 0 with value: 450.0.
[I 2022-09-14 14:22:17,182] Trial 0 finished with value: 450.0 and parameters: {'x': -15, 'y': -15}. Best is trial 0 with value: 450.0.
Trial 0 finished with value: 450.0 and parameters: {'x': -15, 'y': -15}. Best is trial 0 with value: 450.0.
trial index is 1
[I 2022-09-14 14:22:17,604] Trial 1 finished with value: 0.04000000000000001 and parameters: {'x': 0.2, 'y': 0}. Best is trial 1 with value: 0.04000000000000001.
Trial 1 finished with value: 0.04000000000000001 and parameters: {'x': 0.2, 'y': 0}. Best is trial 1 with value: 0.04000000000000001.
Traceback (most recent call last):
  File "C:\Users\cuism\Desktop\ex_DOEs\HF2_code_DOE_enqueue.py", line 42, in <module>
    study.optimize(objective, n_trials=10, n_jobs=2)
  File "C:\Python\Python39\lib\site-packages\optuna\study\study.py", line 419, in optimize
    _optimize(
  File "C:\Python\Python39\lib\site-packages\optuna\study\_optimize.py", line 103, in _optimize
    f.result()
  File "C:\Python\Python39\lib\concurrent\futures\_base.py", line 438, in result
    return self.__get_result()
  File "C:\Python\Python39\lib\concurrent\futures\_base.py", line 390, in __get_result
    raise self._exception
  File "C:\Python\Python39\lib\concurrent\futures\thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "C:\Python\Python39\lib\site-packages\optuna\study\_optimize.py", line 160, in _optimize_sequential
    frozen_trial = _run_trial(study, func, catch)
  File "C:\Python\Python39\lib\site-packages\optuna\study\_optimize.py", line 208, in _run_trial
    frozen_trial = _tell_with_warning(
  File "C:\Python\Python39\lib\site-packages\optuna\study\_tell.py", line 197, in _tell_with_warning
    study._storage.set_trial_state_values(trial_id, state, values)
  File "C:\Python\Python39\lib\site-packages\optuna\storages\_cached_storage.py", line 261, in set_trial_state_values
    self._check_trial_is_updatable(cached_trial)
  File "C:\Python\Python39\lib\site-packages\optuna\storages\_cached_storage.py", line 386, in _check_trial_is_updatable
    raise RuntimeError(
RuntimeError: Trial#0 has already finished and can not be updated.
```

### Steps to reproduce

1. run the attached codes several times, the issue will come
2.
3.
```python
import optuna
import time
import random
import logging
import sys
import os

def objective(trial):
    print("trial index is", trial.number)
   
    x = trial.suggest_float("x", -100, 100)
    y = trial.suggest_float("y", -15, 30)
    obj=x ** 2 + y ** 2
    #time.sleep(random.uniform(1,5))
    return obj
    

file_path = "C:/Users/cuism/Desktop/ex_DOEs/study.db"
if os.path.isfile(file_path):
  os.remove(file_path)
  print("File has been deleted")
else:
  print("File does not exist")
  
optuna.logging.get_logger("optuna").addHandler(logging.StreamHandler(sys.stdout))
study_name = "C:/Users/cuism/Desktop/ex_DOEs/study"  # Unique identifier of the study.
storage_name = "sqlite:///{}.db".format(study_name)  


try_these_parameters_first=[{"x": -15, "y":-15},{"x": 0.2, "y":0},{"x": 30, "y":30},{"x": 20, "y":20}]

sampler = optuna.integration.BoTorchSampler( n_startup_trials=5)
study = optuna.create_study(sampler=sampler,study_name=study_name, storage=storage_name)

study.enqueue_trial(try_these_parameters_first[0])
study.enqueue_trial(try_these_parameters_first[1])
study.enqueue_trial(try_these_parameters_first[2])
study.enqueue_trial(try_these_parameters_first[3])


study.optimize(objective, n_trials=10, n_jobs=2)
```

### Additional context (optional)

_No response_

### PR DESCRIPTION
**Pull request by c-bata**
## Motivation

Close #4002

## Description of the changes

Add an FAQ section and a note in the docstring of RDBStorage.

Screenshot 2022-09-30 14 27 08

https://output.circle-artifacts.com/output/job/a18acc0a-09ce-4506-8ebf-355c2d0b4357/artifacts/0/docs/build/html/faq.html#sqlite-concurrency

Screenshot 2022-09-30 14 26 48

https://output.circle-artifacts.com/output/job/a18acc0a-09ce-4506-8ebf-355c2d0b4357/artifacts/0/docs/build/html/reference/generated/optuna.storages.RDBStorage.html#optuna.storages.RDBStorage

### CODE REVIEW

**Comment by HideakiImamura** [comment]
@knshnb @contramundum53 Could you review this PR?

**Review by contramundum53 — APPROVED** [review]
LGTM!

File: docs/source/faq.rst [hunk 1]
```diff
@@ -537,6 +537,32 @@ However, if you are in the environment where you can not install a client/server
...
+How can I solve the error that occurs when performing parallel optimization with SQLite3?
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We would never recommend SQLite3 for parallel optimization in following reasons.
+
+- To concurrently evaluate trials enqueued by :func:`~optuna.study.Study.enqueue_trial`, :class:`~optuna.storages.RDBStorage` uses `SELECT ... FOR UPDATE` query.
+  However, as you can see `here <https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_4_41/lib/sqlalchemy/dialects/sqlite/base.py#L1265-L1267>`_, it doesn't work with SQLite3 since there is no low-level locking support.
+- Due to the restriction of SQLite3 and Python DB-API, As described in `the documentation of SQLAlchemy <https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#sqlite-concurrency>`_,
+  SQLite3 (and pysqlite driver) is not good at handling high-level of concurrency. It may raise "database is locked" error.
+- For distributed optimization via NFS, SQLite3 does not work since it depends on `fcntl()`, which is broken on many NFS implementations.
```
Comment (c-bata) [inline docs/source/faq.rst:551]: Fix a typo (low-level locking) and improve the description.

```suggestion
- To concurrently evaluate trials enqueued by :func:`~optuna.study.Study.enqueue_trial`, :class:`~optuna.storages.RDBStorage` uses `SELECT ... FOR UPDATE` syntax, which is unsupported in `SQLite3 <https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_4_41/lib/sqlalchemy/dialects/sqlite/base.py#L1265-L1267>`_.
- As described in `the SQLAlchemy's documentation <https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#sqlite-concurrency>`_,
  SQLite3 (and pysqlite driver) does not support a high level of concurrency.
  You may get a "database is locked" error, which occurs when one thread or process has an exclusive lock on a database connection (in reality a file handle) and another thread times out waiting for the lock to be released.
  You can increase the default `timeout <https://docs.python.org/3/library/sqlite3.html#sqlite3.connect>`_ value like `optuna.storages.RDBStorage("sqlite:///example.db", engine_kwargs={"connect_args": {"timeout": 20.0}})` though.
- For distributed optimization via NFS, SQLite3 does not work as described at `here <https://www.sqlite.org/faq.html#q5>`_.
```

**Review by knshnb — APPROVED** [review]
Thanks for the update! Let me merge this PR.

**Pull request merged**